api/src/main/java/org/openmrs/module/registrationcore/api/impl/PatientCreatedListener.java

```diff
@@ -0,0 +1,129 @@
+package org.openmrs.module.registrationcore.api.impl;
+
+import javax.jms.Message;
+import org.apache.commons.lang.exception.ExceptionUtils;
+import org.openmrs.Patient;
+import org.openmrs.PatientIdentifier;
+import org.openmrs.PatientIdentifierType;
+import org.openmrs.User;
+import org.openmrs.event.Event;
+import org.openmrs.module.registrationcore.api.errorhandling.ErrorHandlingService;
+import org.openmrs.module.registrationcore.api.errorhandling.PixErrorHandlingService;
+import org.openmrs.module.registrationcore.api.mpi.common.MpiException;
+import org.openmrs.module.registrationcore.api.mpi.common.MpiProperties;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * This class listens for patient CREATED events.
+ * If MPI is enabled, it exports patient to MPI,
+ * and updates the local patient with the new patient identifier generated by the MPI.
+ */
+public class PatientCreatedListener extends PatientActionListener {
+    
+    private static final Logger LOGGER = LoggerFactory.getLogger(PatientCreatedListener.class);
+    
+    private MpiProperties mpiProperties;
+
+    private IdentifierBuilder identifierBuilder;
+
+    public void setMpiProperties(MpiProperties mpiProperties) {
+        this.mpiProperties = mpiProperties;
+    }
+
+    public void setIdentifierBuilder(IdentifierBuilder identifierBuilder) {
+        this.identifierBuilder = identifierBuilder;
+    }
+
+    /**
+     * Defines the list of Actions that this subscribable event listener class listens out for.
+     *
+     * @return a list of Actions this listener can deal with
+     */
+    public List<String> subscribeToActions(){
+        List actions = new ArrayList<String>();
+        actions.add(Event.Action.CREATED.name());
+        return actions;
+    }
+
+    /**
+     * Exports patient to MPI, and updates the local patient
+     * with the new patient identifier generated by the MPI.
+     *
+     * @param message message with properties.
+     */
+    @Override
+    public void performMpiAction(Message message) {
+        Patient patient = extractPatient(message);
+        if (patient != null){
+            PatientIdentifierType mpiPatientIdType = patientService.getPatientIdentifierTypeByUuid(mpiProperties.getMpiPersonIdentifierTypeUuid());
+            PatientIdentifier mpiPatientId = patient.getPatientIdentifier(mpiPatientIdType);
+            if (mpiPatientId != null && !mpiPatientId.getIdentifier().isEmpty()) {
+                // Patient originated in MPI and thus should not be pushed to the MPI
```

Comment:
I think this comment is a little deceptive, I would think when you get here it implies the patient exists in the MPI and not that they originated, they could have originated in this instance and pushed to MPI.

Reply:
true, I will change this. Also I will add some code in here to fix the bug "When MPI patient without Local Required Identifier is imported, Local ID is not updated in MPI upon generation" which is in the MPI project in the PIH Jira.